jax/lax_control_flow.py

```diff
@@ -0,0 +1,388 @@
+  return maybe_tracer_tuple_to_abstract_tuple(aval_out)
+
+def _while_loop_translation_rule(c, init_val, cond_tracer_consts,
+                                 body_tracer_consts, cond_consts, body_consts,
+                                 aval_out, cond_jaxpr, body_jaxpr):
+  loop_carry = c.Tuple(init_val, cond_tracer_consts, body_tracer_consts)
+  shape = c.GetShape(loop_carry)
+
+  loop_carry_var = pe.Var(0, "loop_carry")
+  outvar = pe.Var(0, "loop_carry_out")
+  cond_var = pe.Var(0, "cond_consts")
+  body_var = pe.Var(0, "body_consts")
+
+  num_cond_consts = len(cond_consts.val)
+  assert len(cond_jaxpr.invars) == 1
+  cond_jaxpr_converted = cond_jaxpr.copy()
+  cond_jaxpr_converted.constvars = cond_jaxpr.constvars[:num_cond_consts]
+  cond_jaxpr_converted.invars = [loop_carry_var]
+  cond_jaxpr_converted.eqns = (
+      [_unpack_eqn(loop_carry_var, [cond_jaxpr.invars[0], cond_var, body_var]),
+       _unpack_eqn(cond_var, cond_jaxpr.constvars[num_cond_consts:])]
+      + list(cond_jaxpr.eqns))
+
+  num_body_consts = len(body_consts.val)
+  assert len(body_jaxpr.invars) == 1
+  body_jaxpr_converted = body_jaxpr.copy()
+  body_jaxpr_converted.constvars = body_jaxpr.constvars[:num_body_consts]
+  body_jaxpr_converted.invars = [loop_carry_var]
+  body_jaxpr_converted.outvar = outvar
+  body_jaxpr_converted.eqns = (
+      [_unpack_eqn(loop_carry_var, [body_jaxpr.invars[0], cond_var, body_var]),
+       _unpack_eqn(body_var, body_jaxpr.constvars[num_body_consts:])]
+      + list(body_jaxpr.eqns) +
+      [_pack_eqn([body_jaxpr.outvar, cond_var, body_var], outvar)])
+
+  cond_computation = xla.jaxpr_computation(
+      cond_jaxpr_converted, cond_consts.val, (), shape)
+  body_computation = xla.jaxpr_computation(
+      body_jaxpr_converted, body_consts.val, (), shape)
+  full_ans = c.While(cond_computation, body_computation, loop_carry)
+  return c.GetTupleElement(full_ans, 0)
+
+def _while_loop_batching_rule(batched_args, batch_dims, cond_consts,
+                              body_consts, aval_out, cond_jaxpr, body_jaxpr):
+  # See https://github.com/google/jax/issues/441 for a discussion.
+  # To batch a while_loop, we need to do some masking, since the elements of the
+  # batch may run for different numbers of iterations. We perform that masking
+  # using lax.select, and keep the loop running so long as any of the batch
+  # elements need by effectively using an np.any(...) in the cond_fun.
+  # The basic strategy here is to lift `cond_jaxpr` and `body_jaxpr` back into
+  # traceable Python functions using `core.eval_jaxpr`. Then we can batch them
+  # using `batching.batch_transform` (the transform underlying `api.vmap`). This
+  # code also avoids broadcasting `cond_tracer_consts` and `body_tracer_consts`.
+  init_val, cond_tracer_consts, body_tracer_consts = batched_args
+  init_val_bd, cond_tracer_consts_bd, body_tracer_consts_bd = batch_dims
+
+  sizes = lax._reduce(set.union, map(batching.dimsize, batch_dims, batched_args))
```

Comment:
This is just the builtin reduce (we rename it to `_reduce` in the lax module namespace because we want to define a function modeling the XLA HLO reduce). We could just call `reduce` here, though we'll also need a `from six.moves import reduce`.